orangecontrib/spectroscopy/widgets/owspectra.py

```diff
@@ -1009,6 +1062,101 @@ def grid_apply(self):
         self.plot.showGrid(self.show_grid, self.show_grid, alpha=0.3)
         self.show_grid_a.setChecked(self.show_grid)
 
+    def find_peak_variables(self):
+        if self.maximum_point is None:
+            max_value = []
+            data = np.array(self.data)
+            for i, val in enumerate(data):
+                single_spcetra = data[i]
+                peaks, _ = find_peaks(single_spcetra)
+                max_value.append(np.amax(single_spcetra[peaks]))
+                # method for finding maxes without interference from extra data labels
+            self.maximum_point = np.amax(max_value)
+        if self.minimum_point is None:
+            self.minimum_point = np.amin(self.data)
+        if self.start_point is None:
+            self.start_point = round(np.median(getx(self.data)), 2)
+
+    def peak_apply(self):
+        if self.viewtype == INDIVIDUAL:
+            Label_line = pl.PeakLine()
+            Label_line.setMovable(True)
+            Label_line.setPos(self.start_point)
+            Label_line.label.setText(str(round(self.start_point), 3))
+            Label_line.setPen(pg.mkPen(color=QColor(Qt.black), width=2, style=Qt.DotLine))
+            Label_line.setSpan(mn=self.minimum_point, mx=self.maximum_point)
+            Label_line.label.setColor(color=QColor(Qt.black))
+            Label_line.label.setPosition(1)
+            Label_line.label.setMovable(True)
+            self.plot.addItem(Label_line)
+            Label_line.updateLabel()
+
+    def peak_apply_auto(self, prominence, minHeight, maxHeight, line_overlap):
+        x_axis = getx(self.data)
+        data = np.array(self.data)
+        peak = []
+        if np.shape(data) != (len(data),):
+            for i, val in enumerate(data):
```

Comment:
Did you perhaps try your code on data with 100000 spectra? Please do.

I am just guessing, but grouping of same peaks will need to be written much more efficiently. Perhaps you could sort the list and then go through it only once. You algorithm complexity is O(n**2) where n is number of all peaks in the whole data set, which will not fly. It needs to be at most O(nlogn), achievable with sorting + inspection of the sorted list. There are probably better ways to do it as I thought of now, though.

Also, if there is a huge data set, looking at its subset of spectra for peaks will probably be sufficient.